### DISCUSSION TITLE
is there a way to provide services which are already wrapped?

### DISCUSSION THREAD

**Discussion by jondot** [actix/actix-web#3176] [Q&A] [answered]
The builder has this:

```rust
    pub fn service<F>(mut self, factory: F) -> Self
    where
        F: HttpServiceFactory + 'static,
    {
        self.services
            .push(Box::new(ServiceFactoryWrapper::new(factory)));
        self
    }
```

However I would like to create my own "controller" abstraction, in which a Vec of `box`ed services is provided (I will do the wrapping and boxing myself). Then I want to call `app.add(<wrapped services>)`

Right now the `services` field is private and there no option to initialize it.

Any idea?

**Comment by robjtede** [comment, marked as answer]
There's no mechanism to allow this, currently. I'd look at an issue or PR showing how you might achieve this if you want to have a stab at it.
